### PR TITLE
Turn status code 429 into warning instead of failure

### PR DESCRIPTION
**Pull request by weshaggard**
Fixes https://github.com/linkchecker/linkchecker/issues/347

As I started looking into I decided to just turn 429 into a warning because the Retry-After was not even provided by some services such as npmjs and in others there were large times like an hour or day so I decided it best to just treat these as a warning instead of a error which will help with my use cases.

### CODE REVIEW

**Comment by anarcat** [comment]
kicking CI again.

**Comment by mgedmin** [comment]
Looks like some tests expect the old behaviour and need to be updated.

**Comment by weshaggard** [comment]
I updated the failing test to handle the new change.

**Comment by anarcat** [comment]
the spec says nothing about crawlers:

https://tools.ietf.org/html/rfc6585#section-4

it says:

>    The 429 status code indicates that the user has sent too many
>    requests in a given amount of time ("rate limiting").
>
>   The response representations SHOULD include details explaining the
>   condition, and MAY include a Retry-After header indicating how long
>   to wait before making a new request.
>   [...]
>    Responses with the 429 status code MUST NOT be stored by a cache.

so i guess treating those as warnings is okay? i'm not sure at all...

 but warnings do show up prominently in linkchecker, maybe that's enough?

**Comment by weshaggard** [comment]
Warning are definitely enough for my scenario but I'll leave it up to you guys whether or not it is the best thing for the overall project. 

FWIW I ended up finding another solution for my link checking.

**Comment by anarcat** [comment]
> FWIW I ended up finding another solution for my link checking.

would be curious to hear what that is! :)

**Comment by weshaggard** [comment]
>would be curious to hear what that is! :)

I wrote a powershell script https://github.com/Azure/azure-sdk/blob/master/eng/scripts/Verify-Links.ps1. I'm certain it isn't perfect but it is relatively simple and serves my needs.

**Comment by weshaggard** [comment]
Closing this PR as I don't need it any longer. If someone else needs it feel free to pull the changes.

**Comment by vincentbernat** [comment]
GitHub started to be quite aggressive around this. It seems the code in this PR is OK. Could it be merged?

**Comment by mgedmin** [comment]
Sure, why not?

**Comment by mgedmin** [comment]
Resolved the merge conflcts, let's see if the CI passes.

I'd love to see a screenshot of what the warning looks like in actual life.  If it looks like the other warnings I've seen recently (yellow text and everything, count at the bottom), then I no longer have any misgivings about this causing any missed problems.

(Which just goes to show that I'm insufficiently familiar with linkchecker to be reviewing pull requests!)

**Comment by mgedmin** [comment]
Ooh I can test it on http://httpbin.org/status/429.

**Comment by mgedmin** [comment]
I've tried to test this, and I smacked right into bug #411:

```
mg@blynas: ~/src/linkchecker [git:master $>] $ linkchecker http://httpbin.org/status/429
INFO linkcheck.cmdline 2020-05-23 14:19:44,080 MainThread Checking intern URLs only; use --check-extern to check extern URLs.
LinkChecker 10.0.0.dev0              Copyright (C) 2000-2014 Bastian Kleineidam
LinkChecker comes with ABSOLUTELY NO WARRANTY!
This is free software, and you are welcome to redistribute it
under certain conditions. Look at the file `LICENSE' within this
distribution.
Get the newest version at https://linkchecker.github.io/linkchecker/
Write comments and bugs to https://github.com/linkchecker/linkchecker/issues

Start checking at 2020-05-23 14:19:44+003


********** Oops, I did it again. *************

You have found an internal error in LinkChecker. Please write a bug report
at https://github.com/linkchecker/linkchecker/issues
and include the following information:
- the URL or file you are testing
- the system information below

When using the commandline client:
- your commandline arguments and any custom configuration files.
- the output of a debug run with option "-Dall"

Not disclosing some of the information above due to privacy reasons is ok.
I will try to help you nonetheless, but you have to give me something
I can work with ;) .

Traceback (most recent call last):
  File "/home/mg/src/linkchecker/linkcheck/director/checker.py", line 103, in check_url
    line: self.check_url_data(url_data)
    locals:
      self = <local> <Checker(CheckThread-http://httpbin.org/status/429, started 140368456709888)>
      self.check_url_data = <local> <bound method Checker.check_url_data of <Checker(CheckThread-http://httpbin.org/status/429, started 140368456709888)>>
      url_data = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
  File "/home/mg/src/linkchecker/linkcheck/director/checker.py", line 115, in check_url_data
    line: check_url(url_data, self.logger)
    locals:
      check_url = <global> <function check_url at 0x7faa18abd040>
      url_data = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self = <local> <Checker(CheckThread-http://httpbin.org/status/429, started 140368456709888)>
      self.logger = <local> <linkcheck.director.logger.Logger object at 0x7faa18a04d00>
  File "/home/mg/src/linkchecker/linkcheck/director/checker.py", line 52, in check_url
    line: do_parse = url_data.check_content()
    locals:
      do_parse = <not found>
      url_data = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      url_data.check_content = <local> <bound method UrlBase.check_content of <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>>
  File "/home/mg/src/linkchecker/linkcheck/checker/urlbase.py", line 468, in check_content
    line: if self.allows_recursion():
    locals:
      self = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self.allows_recursion = <local> <bound method UrlBase.allows_recursion of <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>>
  File "/home/mg/src/linkchecker/linkcheck/checker/urlbase.py", line 557, in allows_recursion
    line: if not self.content_allows_robots():
    locals:
      self = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self.content_allows_robots = <local> <bound method HttpUrl.content_allows_robots of <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>>
  File "/home/mg/src/linkchecker/linkcheck/checker/httpurl.py", line 88, in content_allows_robots
    line: soup = self.get_soup()
    locals:
      soup = <not found>
      self = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self.get_soup = <local> <bound method UrlBase.get_soup of <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>>
  File "/home/mg/src/linkchecker/linkcheck/checker/urlbase.py", line 623, in get_soup
    line: self.get_content()
    locals:
      self = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self.get_content = <local> <bound method HttpUrl.get_content of <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>>
  File "/home/mg/src/linkchecker/linkcheck/checker/httpurl.py", line 310, in get_content
    line: self.text = self.data.decode(self.soup.original_encoding)
    locals:
      self = <local> <http link, base_url='http://httpbin.org/status/429', parent_url=None, base_ref=None, recursion_level=0, url_connection=None, line=None, column=None, page=0, name='', anchor='', cache_url=http://httpbin.org/status/429>
      self.text = <local> None
      self.data = <local> b''
      self.data.decode = <local> <built-in method decode of bytes object at 0x7faa19f4e5a0>
      self.soup = <local> 
      self.soup.original_encoding = <local> None
TypeError: decode() argument 'encoding' must be str, not None
System info:
LinkChecker 10.0.0.dev0
Released on: xx.xx.xxxx
Python 3.8.2 (default, Apr 27 2020, 15:53:34) 
[GCC 9.3.0] on linux
PYTHONSTARTUP = '/home/mg/.python'
Modules: Requests, Pysqlite, Sqlite
Local time: 2020-05-23 14:19:44+003
sys.argv: ['/home/mg/.local/bin/linkchecker', 'http://httpbin.org/status/429']
LC_CTYPE = 'lt_LT.UTF-8'
LANG = 'lt_LT.UTF-8'
Default locale: ('lt', 'UTF-8')

 ******** LinkChecker internal error, over and out ********
WARNING linkcheck.check 2020-05-23 14:19:44,694 CheckThread-http://httpbin.org/status/429 internal error occurred

Statistics:
Downloaded: 0B.
No statistics available since no URLs were checked.

That's it. 0 links in 1 URL checked. 0 warnings found. 0 errors found.
There was 1 internal error.
Stopped checking at 2020-05-23 14:19:45+003 (1 seconds)
```

Let's merge that one first.

**Comment by mgedmin** [comment]
Okay, I've created a test branch based on master and merged this PR and #411 into it.  Here's the result:

<details>

```
mg@blynas: ~/src/linkchecker [git:test $] $ linkchecker http://httpbin.org/status/429
INFO linkcheck.cmdline 2020-05-23 21:12:52,196 MainThread Checking intern URLs only; use --check-extern to check extern URLs.
LinkChecker 10.0.0.dev0              Copyright (C) 2000-2014 Bastian Kleineidam
LinkChecker comes with ABSOLUTELY NO WARRANTY!
This is free software, and you are welcome to redistribute it
under certain conditions. Look at the file `LICENSE' within this
distribution.
Get the newest version at https://linkchecker.github.io/linkchecker/
Write comments and bugs to https://github.com/linkchecker/linkchecker/issues

Start checking at 2020-05-23 21:12:52+003

URL        `http://httpbin.org/status/429'
Real URL   http://httpbin.org/status/429
Check time 0.486 seconds
D/L time   0.000 seconds
Size       0B
Warning    [url-rate-limited] Rate limited (Retry-After: None)
           [url-content-size-zero] Content size is zero.
Result     Valid: 429 TOO MANY REQUESTS

Statistics:
Downloaded: 0B.
Content types: 0 image, 1 text, 0 video, 0 audio, 0 application, 0 mail and 0 other.
URL lengths: min=29, max=29, avg=29.

That's it. 1 link in 1 URL checked. 2 warnings found. 0 errors found.
Stopped checking at 2020-05-23 21:12:53+003 (1 seconds)
```

</details>

And since I wanted a screenshot
Ekrano nuotrauka iš 2020-05-23 21-14-08

It's noticeable enough.

I'm not too happy about the "Valid: 429 TOO MANY REQUESTS" bit, but I'm not sure what other options there are.

**Comment by mgedmin** [comment]
#411 is merged, this one works, do we land it?

@anarcat, @cjmayo: any objections?

**Comment by anarcat** [comment]
no obj.

File: linkcheck/checker/httpurl.py [hunk 1]
```diff
@@ -285,14 +285,19 @@ def getheader(self, name, default=None):
...
             self.set_result("%d %s" % (self.url_connection.status_code, self.url_connection.reason),
                             valid=False)
         else:
             if self.url_connection.status_code == 204:
                 # no content
                 self.add_warning(self.url_connection.reason,
                                  tag=WARN_HTTP_EMPTY_CONTENT)
+
+            if self.url_connection.status_code == 429:
+                self.add_warning(u"Rate limited (Retry-After: %s)" % self.getheader(_("Retry-After")),
```
Comment (cjmayo) [inline linkcheck/checker/httpurl.py:298]: Let's not put u string-prefixes back into the code.

**Pull request merged**